### PR TITLE
Add usrSCTP support

### PR DESCRIPTION
**Pull request by ruengeler**

### CODE REVIEW

**Comment by tuexen** [comment]
Hmm. I would expect that all the neat examples just work. Without anything special related the the usrsctp stack. This should be covered completely within the neat library.

File: neat_core.c [hunk 1]
```diff
@@ -222,7 +229,7 @@ static void free_cb(uv_handle_t *handle)
     flow->closefx(flow->ctx, flow);
     free((char *)flow->name);
     free((char *)flow->port);
-   if (flow->resolver_results) {
+    if (flow->resolver_results) {
```
Comment (tuexen) [inline neat_core.c]: Whitespace change?

File: neat_core.c [hunk 2]
```diff
@@ -434,7 +451,12 @@ he_connected_cb(uv_poll_t *handle, int status, int events)
         flow->sockType = he_ctx->candidate->ai_socktype;
         flow->sockProtocol = he_ctx->candidate->ai_protocol;
         flow->everConnected = 1;
+#if defined(USRSCTP_SUPPORT)
+        flow->sock = he_ctx->sock;
+#else
         flow->fd = he_ctx->fd;
+#endif
+
```
Comment (tuexen) [inline neat_core.c]: I'm not sure here, but don't you assume that if USRSCTP_SUPPORT is defined all sockets are handled via the library? This is not true, since you would still use kernel sockets for TCP, UDP and UDPLite. I guess you have to check that.

File: neat_core.c [hunk 3]
```diff
@@ -547,14 +576,25 @@ static void do_accept(neat_ctx *ctx, neat_flow *flow)
 
     newFlow->handle = (uv_poll_t *) malloc(sizeof(uv_poll_t));
     assert(newFlow->handle != NULL);
+
+#if defined(USRSCTP_SUPPORT)
+    newFlow->sock = newFlow->acceptfx(ctx, newFlow, flow->sock);
+    if (!newFlow->sock) {
+#else
```
Comment (tuexen) [inline neat_core.c]: Same comment as above: Does you code handle TCP, UDP and UDPLite kernel sockets when using usrsctp?

File: neat_core.c [hunk 4]
```diff
@@ -730,7 +774,14 @@ neat_write_via_kernel_flush(struct neat_ctx *ctx, struct neat_flow *flow)
             msghdr.msg_controllen = 0;
 #endif
             msghdr.msg_flags = 0;
+#if !defined(USRSCTP_SUPPORT)
             rv = sendmsg(flow->fd, (const struct msghdr *)&msghdr, 0);
+#else
```
Comment (tuexen) [inline neat_core.c]: Same question as above.

File: neat_core.c [hunk 5]
```diff
@@ -822,7 +873,7 @@ neat_write_via_kernel(struct neat_ctx *ctx, struct neat_flow *flow,
     struct iovec iov;
 #if defined(SCTP_SNDINFO)
     char cmsgbuf[CMSG_SPACE(sizeof(struct sctp_sndinfo))];
-    struct sctp_sndinfo *sndinfo;
+    struct sctp_sndinfo *sndinfo = NULL;
```
Comment (tuexen) [inline neat_core.c]: Why is that needed?

File: neat_core.c [hunk 6]
```diff
@@ -1103,6 +1163,184 @@ neat_shutdown_via_kernel(struct neat_ctx *ctx, struct neat_flow *flow)
...
+        flow->closefx(he_ctx->nc, flow);
+        free(he_ctx);
+    }
+    return 0;
+}
+
+static int
+neat_close_via_usrsctp(struct neat_ctx *ctx, struct neat_flow *flow)
+{
+    printf("neat_close_via_usrsctp\n");
```
Comment (tuexen) [inline neat_core.c]: Don't use printf()s in the library code. Maybe you can use the new logging stuff.

File: neat_core.c [hunk 7]
```diff
@@ -1103,6 +1163,184 @@ neat_shutdown_via_kernel(struct neat_ctx *ctx, struct neat_flow *flow)
...
+}
+
+static int
+neat_listen_via_usrsctp(struct neat_ctx *ctx, struct neat_flow *flow)
+{
+    int enable = 1;
+    socklen_t len;
+    int size;
+
+    printf("neat_listen_via_usrsctp\n");
```
Comment (tuexen) [inline neat_core.c]: See above.

File: neat_core.c [hunk 8]
```diff
@@ -1132,15 +1370,26 @@ neat_flow *neat_new_flow(neat_ctx *mgr)
...
         return NULL;
 
-    rv->fd = -1;
     rv->handle = NULL;
-    rv->writefx = neat_write_via_kernel;
-    rv->readfx = neat_read_via_kernel;
+    rv->writefx = neat_write_to_lower_layer;
+    rv->readfx = neat_read_from_lower_layer;
+#if !defined(USRSCTP_SUPPORT)
+    rv->fd = -1;
```
Comment (tuexen) [inline neat_core.c]: I guess this is the issue with supporting kernel sockets in addition to the usrsctp stack.

File: neat_he.c [hunk 9]
```diff
@@ -136,8 +136,11 @@ he_resolve_cb(struct neat_resolver *resolver, struct neat_resolver_results *resu
         he_ctx->nc = resolver->nc;
         he_ctx->candidate = candidate;
         he_ctx->flow = flow;
+#ifdef USRSCTP_SUPPORT
+        he_ctx->sock = NULL;
+#else
         he_ctx->fd = -1;
-
+#endif
```
Comment (tuexen) [inline neat_he.c]: See above.

File: samples/neat_basic_usrsctp.c [hunk 10]
```diff
@@ -0,0 +1,132 @@
+#include <stdlib.h>
```
Comment (tuexen) [inline samples/neat_basic_usrsctp.c]: I think we should integrate usrsctp such that no specific examples are needed.

File: neat_core.c (hunk 6 above)
Comment (weinrank) [inline neat_core.c]: I will take care of it

File: neat_he.c [hunk 11]
```diff
@@ -151,7 +154,8 @@ he_resolve_cb(struct neat_resolver *resolver, struct neat_resolver_results *resu
         if (flow->connectfx(he_ctx, callback_fx) == -1) {
             /* TODO: Some error handling? */
             continue;
-        }
+        } else
+            break;
```
Comment (weinrank) [inline neat_he.c]: This has to be removed when HE is fixed - sticky note!!! @ruengeler
